taca/utils/undetermined.py

```diff
@@ -0,0 +1,164 @@
+        logger.info("linking file {} to {}".format(fastqfile, path_per_lane[lane]))
+        os.symlink(fastqfile, os.path.join(path_per_lane[lane], os.path.basename(fastqfile)))
+
+def save_index_count(barcodes, run, lane):
+    """writes the barcode counts"""
+    with open(os.path.join(run, dmux_folder, 'index_count_L{}.tsv'.format(lane)), 'w') as f:
+        for barcode in sorted(barcodes, key=barcodes.get, reverse=True):
+            f.write("{}\t{}\n".format(barcode, barcodes[barcode]))
+
+def check_index_freq(run, lane,freq_tresh):
+    """uses subprocess to perform zcat <file> | sed -n '1~4 p' | awk -F ':' '{print $NF}', counts the barcodes and 
+    returns true if the most represented index accounts for less than freq_tresh% of the total"""
+    barcodes={}
+    if os.path.exists(os.path.join(run, dmux_folder,'index_count_L{}.tsv'.format(lane))):
+        logger.info("Found index count for lane {}, skipping.".format(lane))
+        return False
+    else:
+        open(os.path.join(run, dmux_folder,'index_count_L{}.tsv'.format(lane)), 'a').close()
+        for fastqfile in glob.glob(os.path.join(run, dmux_folder, 'Undetermined_*_L00{}_R1*'.format(lane))):
+            logger.info("working on {}".format(fastqfile))
+            zcat=subprocess.Popen(['zcat', fastqfile], stdout=subprocess.PIPE)
+            sed=subprocess.Popen(['sed', '-n', "1~4p"],stdout=subprocess.PIPE, stdin=zcat.stdout)
+            awk=subprocess.Popen(['awk', '-F', ":", '{print $NF}'],stdout=subprocess.PIPE, stdin=sed.stdout)
+            zcat.stdout.close()
+            sed.stdout.close()
+            output = awk.communicate()[0]
+            zcat.wait()
+            sed.wait()
+            for barcode in output.split('\n')[:-1]:
+                try:
+                    barcodes[barcode]=barcodes[barcode]+1
+                except KeyError:
+                    barcodes[barcode]=1
+
+        save_index_count(barcodes, run, lane)
+        total=sum(barcodes.values())
+        count, bar = max((v, k) for k, v in barcodes.items())
+        if total * freq_tresh / 100<count:
+            logger.warn("The most frequent barcode of lane {} ({}) found in {} represents {}%, which is over the threshold of {}%".format(lane, bar, fastqfile, count / total * 100, freq_tresh))
+            return False
+        else:
+            return True
+
+
+
+
+def first_qc_check(lane, lb, und_tresh, q30_tresh):
+    """checks wether the percentage of bases over q30 for the sample is above the treshold, and if the amount of undetermined is below the treshold"""
+    d={}
+    for entry in lb.sample_data:
+        if lane == int(entry['Lane']):
+            if entry.get('Sample')=='unknown':
+                if float(entry['% >= Q30bases']) < q30_tresh:
+                    logger.warn("Undetermined indexes of lane {} has a percentage of bases over q30 of {}%, which is below the cutoff of {}% ".format(lane, float(entry['% >= Q30bases']), q30_tresh))
```

Comment:
Why is there a separate check for undetermined %Q30? Isn't the point to be within Illuminas specifications? Like %Q30 >= 75 for 2x150 for the entire lane.